File: src/main/java/org/tobi29/scapes/engine/gui/Gui.java
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.gui;

import java8.util.Optional;
import java8.util.concurrent.ConcurrentMaps;
import java8.util.function.Consumer;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.FastMath;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public abstract class Gui extends GuiComponentSlabHeavy {
    protected final GuiStyle style;
    private final List<SelectionEntry> selections = new ArrayList<>();
    private final ConcurrentMap<GuiAction, Set<Consumer<ScapesEngine>>>
            actions = new ConcurrentHashMap<>();
    private GuiComponent lastClicked;
    private int selection = -1, selectionColumn;

    protected Gui(GuiStyle style) {
        super(new GuiLayoutDataRoot());
        this.style = style;

        on(GuiAction.ACTIVATE, engine -> {
            synchronized (selections) {
                if (selection < 0) {
                    return;
                }
                SelectionEntry entry = selections.get(selection);
                sendNewEvent(GuiEvent.CLICK_LEFT, new GuiComponentEvent(),
                        entry.components.get(FastMath.min(selectionColumn,
                                entry.components.size() - 1)), engine);
            }
        });
        on(GuiAction.UP, () -> {
            synchronized (selections) {
                selection = FastMath.max(selection - 1,
                        FastMath.min(0, selections.size() - 1));
            }
        });
        on(GuiAction.DOWN, () -> {
            synchronized (selections) {
                selection = FastMath.min(selection + 1, selections.size() - 1);
            }
        });
        on(GuiAction.LEFT, engine -> {
            synchronized (selections) {
                if (selection < 0) {
                    return;
                }
                SelectionEntry entry = selections.get(selection);
                if (selectionColumn > 0) {
                    selectionColumn = FastMath.min(selectionColumn,
                            entry.components.size() - 1);
                    selectionColumn = FastMath.max(selectionColumn - 1, 0);
                }
                sendNewEvent(GuiEvent.SCROLL,
                        new GuiComponentEvent(Double.NaN, Double.NaN, 1, 0,
                                false), entry.components.get(FastMath
                                .min(selectionColumn,
                                        entry.components.size() - 1)), engine);
            }
        });
        on(GuiAction.RIGHT, engine -> {
            synchronized (selections) {
                if (selection < 0) {
                    return;
                }
                SelectionEntry entry = selections.get(selection);
                if (selectionColumn < entry.components.size() - 1) {
                    selectionColumn = FastMath.min(selectionColumn + 1,
                            entry.components.size() - 1);
                }
                sendNewEvent(GuiEvent.SCROLL,
                        new GuiComponentEvent(Double.NaN, Double.NaN, -1, 0,
                                false), entry.components.get(FastMath
                                .min(selectionColumn,
                                        entry.components.size() - 1)), engine);
            }
        });
    }

    protected void selection(GuiComponent component) {
        selection(component.parent.priority(), component);
    }

    protected void selection(long priority, GuiComponent component) {
        addSelection(priority, Streams.collect(component));
    }

    protected void selection(GuiComponent... components) {
        if (components.length == 0) {
            return;
        }
        selection(components[0].parent.priority(), components);
    }

    protected void selection(long priority, GuiComponent... components) {
        if (components.length == 0) {
            return;
        }
        addSelection(priority, Streams.collect(components));
    }

    protected void selection(List<GuiComponent> components) {
        if (components.isEmpty()) {
            return;
        }
        selection(components.get(0).parent.priority(), components);
    }

    protected void selection(long priority, List<GuiComponent> components) {
        if (components.isEmpty()) {
            return;
        }
        List<GuiComponent> list = new ArrayList<>();
        list.addAll(components);
        addSelection(priority, list);
    }

    private void addSelection(long priority, List<GuiComponent> components) {
        SelectionEntry entry = new SelectionEntry(priority, components);
        synchronized (selections) {
            for (int i = selections.size() - 1; i >= 0; i--) {
                if (selections.get(i).priority >= priority) {
                    selections.add(i + 1, entry);
                    return;
                }
            }
            selections.add(0, entry);
        }
    }

    public void on(GuiAction action, Runnable listener) {
        on(action, engine -> listener.run());
    }

    public void on(GuiAction action, Consumer<ScapesEngine> listener) {
        Set<Consumer<ScapesEngine>> listeners = ConcurrentMaps
                .computeIfAbsent(actions, action, key -> Collections
                        .newSetFromMap(new ConcurrentHashMap<>()));
        listeners.add(listener);
    }

    public Optional<GuiComponent> fireNewEvent(GuiEvent type,
            GuiComponentEvent event, ScapesEngine engine) {
        return fireNewEvent(event, GuiComponent.sink(type), engine);
    }

    public Optional<GuiComponent> fireNewEvent(GuiComponentEvent event,
            EventSink listener, ScapesEngine engine) {
        return fireEvent(new GuiComponentEvent(event, baseSize(engine)),
                listener, engine);
    }

    public Set<GuiComponent> fireNewRecursiveEvent(GuiEvent type,
            GuiComponentEvent event, ScapesEngine engine) {
        return fireNewRecursiveEvent(event, GuiComponent.sink(type), engine);
    }

    public Set<GuiComponent> fireNewRecursiveEvent(GuiComponentEvent event,
            EventSink listener, ScapesEngine engine) {
        return fireRecursiveEvent(
                new GuiComponentEvent(event, baseSize(engine)), listener,
                engine);
    }

    public boolean sendNewEvent(GuiEvent type, GuiComponentEvent event,
            GuiComponent destination, ScapesEngine engine) {
        return sendNewEvent(event, destination,
                GuiComponent.sink(type, destination), engine);
    }

    public boolean sendNewEvent(GuiComponentEvent event,
            GuiComponent destination, EventDestination listener,
            ScapesEngine engine) {
        return sendEvent(new GuiComponentEvent(event, baseSize(engine)),
                destination, listener, engine);
    }

    public boolean fireAction(GuiAction action, ScapesEngine engine) {
        Set<Consumer<ScapesEngine>> listeners = actions.get(action);
        if (listeners == null || listeners.isEmpty()) {
            return false;
        }
        Streams.forEach(listeners, listener -> listener.accept(engine));
        return true;
    }

    public abstract boolean valid();

    public GuiStyle style() {
        return style;
    }

    public GuiComponent lastClicked() {
        return lastClicked;
    }

    protected void setLastClicked(GuiComponent component) {
        lastClicked = component;
    }

    @Override
    protected void update(ScapesEngine engine, double delta) {
        super.update(engine, delta);
        if (visible) {
            synchronized (selections) {
                Iterator<SelectionEntry> iterator = selections.iterator();
                while (iterator.hasNext()) {
                    SelectionEntry entry = iterator.next();
                    Iterator<GuiComponent> componentIterator =
                            entry.components.iterator();
                    while (componentIterator.hasNext()) {
                        GuiComponent component = componentIterator.next();
                        if (component.removed) {
                            componentIterator.remove();
                        }
                    }
                    if (entry.components.isEmpty()) {
                        iterator.remove();
                        selection =
                                FastMath.min(selection, selections.size() - 1);
                    }
                }
                if (selection < 0) {
                    return;
                }
                SelectionEntry entry = selections.get(selection);
                GuiComponent component = entry.components.get(FastMath
                        .min(selectionColumn, entry.components.size() - 1));
                sendNewEvent(new GuiComponentEvent(), component,
                        component::hover, engine);
            }
        }
    }

    @Override
    public boolean ignoresEvents() {
        return true;
    }

    private static class SelectionEntry {
        public final long priority;
        public final List<GuiComponent> components;

        public SelectionEntry(long priority, List<GuiComponent> components) {
            this.priority = priority;
            this.components = components;
        }
    }
}

File: src/main/kotlin/org/tobi29/scapes/engine/gui/Gui.kt
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.gui

import java8.util.concurrent.ConcurrentMaps
import org.tobi29.scapes.engine.utils.collect
import org.tobi29.scapes.engine.utils.math.max
import org.tobi29.scapes.engine.utils.math.min
import java.util.*
import java.util.concurrent.ConcurrentHashMap

abstract class Gui protected constructor(val style: GuiStyle) : GuiComponentSlabHeavy(
        GuiLayoutDataRoot()) {
    private val selections = ArrayList<SelectionEntry>()
    private val actions = ConcurrentHashMap<GuiAction, MutableSet<() -> Unit>>()
    var lastClicked: GuiComponent? = null
    private var selection = -1
    private var selectionColumn = 0

    init {
        on(GuiAction.ACTIVATE, {
            synchronized(selections) {
                if (selection < 0) {
                    return@on
                }
                val entry = selections[selection]
                sendNewEvent(GuiEvent.CLICK_LEFT, GuiComponentEvent(),
                        entry.components[min(selectionColumn,
                                entry.components.size - 1)])
            }
        })
        on(GuiAction.UP, {
            synchronized(selections) {
                selection = max(selection - 1,
                        min(0, selections.size - 1))
            }
        })
        on(GuiAction.DOWN, {
            synchronized(selections) {
                selection = min(selection + 1, selections.size - 1)
            }
        })
        on(GuiAction.LEFT, {
            synchronized(selections) {
                if (selection < 0) {
                    return@on
                }
                val entry = selections[selection]
                if (selectionColumn > 0) {
                    selectionColumn = min(selectionColumn,
                            entry.components.size - 1)
                    selectionColumn = max(selectionColumn - 1, 0)
                }
                sendNewEvent(GuiEvent.SCROLL,
                        GuiComponentEvent(Double.NaN,
                                Double.NaN, 1.0, 0.0,
                                false), entry.components[min(selectionColumn,
                        entry.components.size - 1)])
            }
        })
        on(GuiAction.RIGHT, {
            synchronized(selections) {
                if (selection < 0) {
                    return@on
                }
                val entry = selections[selection]
                if (selectionColumn < entry.components.size - 1) {
                    selectionColumn = min(selectionColumn + 1,
                            entry.components.size - 1)
                }
                sendNewEvent(GuiEvent.SCROLL,
                        GuiComponentEvent(Double.NaN,
                                Double.NaN, -1.0, 0.0,
                                false), entry.components[min(selectionColumn,
                        entry.components.size - 1)])
            }
        })
    }

    protected fun selection(component: GuiComponent) {
        selection(component.parent.priority, component)
    }

    protected fun selection(priority: Long,
                            component: GuiComponent) {
        addSelection(priority, collect(component))
    }

    protected fun selection(vararg components: GuiComponent) {
        if (components.size == 0) {
            return
        }
        selection(components[0].parent.priority, *components)
    }

    protected fun selection(priority: Long,
                            vararg components: GuiComponent) {
        if (components.size == 0) {
            return
        }
        addSelection(priority, collect(*components))
    }

    protected fun selection(components: List<GuiComponent>) {
        if (components.isEmpty()) {
            return
        }
        selection(components[0].parent.priority, components)
    }

    protected fun selection(priority: Long,
                            components: List<GuiComponent>) {
        if (components.isEmpty()) {
            return
        }
        val list = ArrayList<GuiComponent>()
        list.addAll(components)
        addSelection(priority, list)
    }

    private fun addSelection(priority: Long,
                             components: MutableList<GuiComponent>) {
        val entry = SelectionEntry(priority, components)
        synchronized(selections) {
            for (i in selections.indices.reversed()) {
                if (selections[i].priority >= priority) {
                    selections.add(i + 1, entry)
                    return
                }
            }
            selections.add(0, entry)
        }
    }

    fun on(action: GuiAction,
           listener: () -> Unit) {
        val listeners = ConcurrentMaps.computeIfAbsent(actions, action) { key ->
            Collections.newSetFromMap(ConcurrentHashMap())
        }
        listeners.add(listener)
    }

    fun fireNewEvent(type: GuiEvent,
                     event: GuiComponentEvent): GuiComponent? {
        return fireNewEvent(event, GuiComponent.sink(type))
    }

    fun fireNewEvent(event: GuiComponentEvent,
                     listener: Function2<GuiComponent, GuiComponentEvent, Boolean>): GuiComponent? {
        return fireEvent(GuiComponentEvent(event, baseSize()), listener)
    }

    fun fireNewRecursiveEvent(type: GuiEvent,
                              event: GuiComponentEvent): Set<GuiComponent> {
        return fireNewRecursiveEvent(event, GuiComponent.sink(type))
    }

    fun fireNewRecursiveEvent(event: GuiComponentEvent,
                              listener: Function2<GuiComponent, GuiComponentEvent, Boolean>): Set<GuiComponent> {
        return fireRecursiveEvent(GuiComponentEvent(event, baseSize()),
                listener)
    }

    fun sendNewEvent(type: GuiEvent,
                     event: GuiComponentEvent,
                     destination: GuiComponent): Boolean {
        return sendNewEvent(event, destination,
                GuiComponent.sink(type, destination))
    }

    fun sendNewEvent(event: GuiComponentEvent,
                     destination: GuiComponent,
                     listener: (GuiComponentEvent) -> Unit): Boolean {
        return sendEvent(GuiComponentEvent(event, baseSize()), destination,
                listener)
    }

    fun fireAction(action: GuiAction): Boolean {
        val listeners = actions[action]
        if (listeners == null || listeners.isEmpty()) {
            return false
        }
        listeners.forEach { it() }
        return true
    }

    abstract val isValid: Boolean

    public override fun update(delta: Double) {
        super.update(delta)
        if (isVisible) {
            synchronized(selections) {
                val iterator = selections.iterator()
                while (iterator.hasNext()) {
                    val entry = iterator.next()
                    val componentIterator = entry.components.iterator()
                    while (componentIterator.hasNext()) {
                        val component = componentIterator.next()
                        if (component.removed) {
                            componentIterator.remove()
                        }
                    }
                    if (entry.components.isEmpty()) {
                        iterator.remove()
                        selection = min(selection, selections.size - 1)
                    }
                }
                if (selection < 0) {
                    return
                }
                val entry = selections[selection]
                val component = entry.components[min(selectionColumn,
                        entry.components.size - 1)]
                sendNewEvent(GuiComponentEvent(), component,
                        { component.hover(it) })
            }
        }
    }

    override fun ignoresEvents(): Boolean {
        return true
    }

    private class SelectionEntry(val priority: Long, val components: MutableList<GuiComponent>)
}
